### ISSUE TITLE
Optimize authentication flows

### ISSUE THREAD

**Issue by cavis** [publicmediaplatform/pmp-issues#47]
It's time to put some speed holes in the PMP.

Currently, most PMP requests require looking up a user by auth-token.  This involves hitting ElasticSearch, and checking for token existence + expiration.  Though it's pretty quick, caching that in something like Redis would be even quicker.  I'll need to do some benchmarks, but it should nearly cut in half the response time for fetch-by-guid.

**Comment by cavis** [comment]
Some performance results, comparing:

* Current master
* Current master with a big bump to `maxSockets` (allow more http connections should speed up ES)
* Optimize auth branch (fetch auth from Postgres, not ES)

![image](https://cloud.githubusercontent.com/assets/1410587/6149197/316d46c8-b1c4-11e4-8f29-0bbdc7973166.png)
![image](https://cloud.githubusercontent.com/assets/1410587/6149236/804b9ef2-b1c4-11e4-87fb-5421324129fb.png)

**Comment by cavis** [comment]
Interestingly, the fetch-from-postgres branch seems to double the throughput of master.  And nearly doubles the performance of the maxSockets change.

I'm also realizing now that I never put an index on the PG column it's reading - so that may get slightly faster.  And redis-caching that entire read would also help.

Just to make sure I wasn't crazy, and Postgres-select-by-guids are faster than ES-filter-by-guids, I ran some simple doc fetch benchmarks.  Both locally (all dependencies on one box), and out in EC2 land (remote PG and ES).

These ES queries are a simple `query: {match: {guid: '1234'}}`

### localhost

type        | avg   | 90th  | 50th  | min   | max   |
----------- | ----- | ----- | ----- | ----- | ----- |
PG-series   | 0.79  | 1     | 1     | 0     | 23    |
ES-series   | 2.74  | 3     | 3     | 2     | 13    |
PG-parallel | 247.9 | 381   | 259   | 40    | 407   |
ES-parallel | 591.5 | 812.5 | 590   | 286   | 874   |

### ec2

type        | avg   | 90th  | 50th  | min   | max   |
----------- | ----- | ----- | ----- | ----- | ----- |
PG-series   | 0.894 | 1     | 1     | 0     | 31    |
ES-series   | 60.24 | 65    | 58    | 53    | 260   |
PG-parallel | 230.4 | 342.5 | 235   | 52    | 368   |
ES-parallel | 1292  | 2024  | 1402  | 568   | 2189  |

**Comment by johnymonster** [comment]
can you reference your test code?  I'm not seeing the same sort of latency with an elastic cluster I have setup here at NPR.

**Comment by cavis** [comment]
Sure -> https://gist.github.com/cavis/cf80d10600dd8e0c4948.  (Notice that I don't close the pg pool until the end - if I close it after each request, the response times are much closer together).

**Comment by johnymonster** [comment]
so this is also related to your maxSockets setting at the top of the gist and to my original point.  You are going to see high load times on the ES side because you are only allotting 30 sockets for http requests in your test code.  the first 30 requests are going to happen concurrently while the remaining N requests wait in an internal http queue.  once a socket is available a connection can be made.

With pg you are opening up a pool of sockets and making no further handshakes, which is why you are seeing a pretty consistent avg request time.

With that being said I would expect your exact results given that the pg lib has an unfair advantage over the restricted socket max.

you should set your maxSockets = 5000 and give it another try.  You should see a dramatic improvement over pg.  I rarely see over 50ms response times from Elastic.

**Comment by johnymonster** [comment]
As for the PG series result, your benefitting from a pool of connections that are already established and thus don't have the connection overhead.  However in a real world scenario you are not always going to maintain that connection pool and will constantly be opening connections to pg, which would level the playing field a little more.  

With the added benefit of graceful degradation of services from an elastic cluster over the Master/slave scenario of PG the ~30ms speed increase over Elastic for having a connection pool becomes mute.

**Comment by cavis** [comment]
Hmm... changing `maxSockets` to 5000 or even Infinity yields just about the same results.  Hmm... (/me puts on troubleshooting cap).

**Comment by johnymonster** [comment]
interesting.  there might be something more going on, maybe try this against your cluster to see if this works any better.

https://gist.github.com/johnymonster/bc820cc08877cddabfaf

you should be able to get all connections relatively consistent as a baseline.  I'm also remembering there is a sweet spot between number of connections and number of requests.  This is due to working with the http api.  each new connection has overhead associated with it.  if every request is a new connection then its going to slow things down for sure so setting maxSockets to 5000 would be too much.  it was quite a while ago that I did the load testing on elastic.  I can't tell 100% if PG uses a compiles library for connections or if its using http.

**Comment by cavis** [comment]
Running my usual chain of `https://elb --> ec2:nginx:basicauth --> localhost:9200`, I get about 100ms to 115ms.  It's only a couple of ms faster going directly to `ec2:9200`.  So at least that auth/failover strategy isn't too costly!

What sort of response times do you get from `maxSockets=100`?  These pings should probably be much faster, if you're saying that actual ES-queries come back < 50ms.  Even running locally, I'm in the 60-100ms range.  Though it gets slightly better if I decrease to about 10 sockets.

**Comment by cavis** [comment]
Reverted my PR to a less-drastic change.  Now it's just using a redis cache in front of searches for users/clients/tokens/groups.  Seems to buy about the same performance gain - plus group updates and credential-changes are real time!

I'm going to deploy this to the Sandbox environment, and see how things shake out there for awhile.  Also planning to try out hosted-elasticsearch there.  (I've tried found.no in a test environment, and got pretty similar results to my beefier Chef-built ES servers).
